src/main/scala/strawman/collection/Map.scala

```diff
@@ -0,0 +1,44 @@
+package strawman.collection
+
+import strawman.collection.mutable.Builder
+
+import scala.Option
+import scala.annotation.unchecked.uncheckedVariance
+import scala.Predef.???
+
+/** Base Map type */
+trait Map[K, +V]
+  extends Iterable[(K, V)]
```

Comment:
It's a design decision on specificity / principle of least power.

The subject here is Map, a top-level trait. Must we constrain its Iterable element type to be Tuple2 or should we allow it instead to return any type of Product2?

Does the correctness of the collections in this rewrite depend on the correctness of the implementation of equals and hashCode of the Iterable element type? If so then there's more safety in using Tuple2 (although not complete safety as Tuple2 isn't final so one could still _technically_ override and subvert its synthetic equals/hashCode).

An example of the benefits in using Product2 instead is that in one's own implementation of Map one could avoid a lot of unboxing and boxing _into_ Tuple2 by using an element type that implements Product2.

One's hand is slightly swayed (possibly unconsciously) by the fact that in Scala, at both the value and type level, the syntax (..) is tuple syntax, instead of product syntax. Therefore requiring the use of Product1/Product2/.. to be used for product syntax, which is more verbose.

Thank you for the question, it allowed me to think out my thoughts and present them in full.